brainiak/utils/fmrisim.py

```diff
@@ -539,56 +573,144 @@ def export_stimfunction(stimfunction,
-                     response_delay=6,
-                     undershoot_delay=12,
-                     response_dispersion=0.9,
-                     undershoot_dispersion=0.9,
-                     response_scale=1,
-                     undershoot_scale=0.035,
-                     scale_function=1,
-                     temporal_resolution=1000.0,
-                     ):
-    """Return a double gamma HRF
+def export_epoch_file(stimfunction,
+                      filename,
+                      tr_duration,
+                      temporal_resolution=1000.0
+                      ):
+    """ Output an epoch file, necessary for some inputs into brainiak
+
+    This takes in the time course of stimulus events and outputs the epoch
+    file used in Brainiak.
 
     Parameters
     ----------
-        stimfunction : list, bool
-            What is the time course of events to be modelled in this
-            experiment
 
-        tr_duration : float
+    stimfunction : list of timepoint by condition arrays
+        The stimulus function describing the time course of events. Each
+        list entry is from a different participant, each row is a different
+        timepoint, each column is a different condition
 
-        response_delay : float
-            How many seconds until the peak of the HRF
+    filename : str
+        The name of the three column text file to be output
 
-        undershoot_delay : float
-            How many seconds until the trough of the HRF
+    tr_duration : float
+        How long is each TR in seconds
 
-        response_dispersion : float
-            How wide is the rising peak dispersion
+    temporal_resolution : float
+        How many elements per second are you modeling with the
+        stimfunction?
 
-        undershoot_dispersion : float
-            How wide is the undershoot dispersion
+    """
 
-        response_scale : float
-             How big is the response relative to the peak
+    # Cycle through the participants, different entries in the list
+    epoch_file = [0] * len(stimfunction)
+    for participant_counter in list(range(0, len(stimfunction))):
 
-        undershoot_scale :float
-            How big is the undershoot relative to the trough
+        # What is the time course for the participant (binarized)
+        stimfunction_ppt = np.abs(stimfunction[participant_counter]) > 0
 
-        scale_function : bool
-            Do you want to scale the function to a range of 1
+        # Cycle through conditions
+        conditions = stimfunction_ppt.shape[1]
+        for condition_counter in list(range(0, conditions)):
 
-        temporal_resolution : float
-            How many elements per second are you modeling for the stim function
-    Returns
+            # Down sample the stim function
+            stride = tr_duration * temporal_resolution
+            stimfunction_temp = stimfunction_ppt[:, condition_counter]
+            stimfunction_temp = stimfunction_temp[::int(stride)]
```

Comment:
This is an important point but I don't have a good answer. I thought about doing averaging but I don't think this is appropriate either. For instance, if the event lasted for even 0.01ms this would suggest that a (small) HRF would be predicted and I am not sure if that is appropriate. You might have other ideas so I am open to changing this.

Reply:
I actually don't think this is a problem. A 0.01 s (I assume you meant s instead of ms) stimulus would predict an HRF almost indistinguishable from that of a 1 s stimulus multiplied by 1/100 because of the slow HRF (you can simulate this with your own functions). Such small a magnitude wouldn't influence GLM result in general. But if a 0.5 s stimulus get undetected, then there may be a problem. This might be left for future due to time pressure. But I highly suggest fixing this.

Reply:
I did infact mean 0.01ms. Since the HRF is scaled if an impulse occurs for any non zero time it will elicit the same HRF as an event lasting about 1s. This is a problem I don't think has any easy solution but I would love to hear if you have ideas

Reply:
Oh, now I see your point. I have always assumed the case where the time series after convolving with HRF is never scaled. In that case there is no problem for 0.01 ms stimulus -- now you see one advantage of not scaling the convolved result. Another example that scaling is not desirable is when the user intentionally sets different weights for events of different conditions. In that case if you scale the signal, you might just undo what the user wants to. And I am not aware of other fMRI packages that do this kind of scaling. Well it is up for your decision. I would suggest not scaling by default, then all the problems are solved.

Reply:
I see your point but I am going to keep it scaled for two reasons: I would need to change all of the example code that assumes the default and the mapping from the stimfunction weights to the convolved response is not obvious. The peak of the unscaled convolved response depends on both the size of the weights and the temporal resolution. For instance doubling the weights in stimfunction doubles the peak of the convolved response but also doubling the temporal resolution also doubles the convolved response. There would be a way around this in which I scale the HRF to 1 and then multiply by the max of stimfunction. This would prevent these kinds of problems but it wouldn't fix the issue raised about brief events. Regardless, I think this is a place where user caution is necessary

Reply:
OK. I think it is fine for now. One note about the dependence on temporal resolution: it can be solved by just always dividing the end result by the temporal resolution number.